sunkit_image/nafe.py

```diff
@@ -0,0 +1,411 @@
+Check LICENSE fore more details.
+
+This module is an implementation of:
+@article{0067-0049-207-2-25,
+  author={M. Druckmüller},
+  title={A Noise Adaptive Fuzzy Equalization Method for Processing Solar
+         Extreme Ultraviolet Images},
+  journal={The Astrophysical Journal Supplement Series},
+  volume={207},
+  number={2},
+  pages={25},
+  url={http://stacks.iop.org/0067-0049/207/i=2/a=25},
+  year={2013},
+  abstract={A new image enhancement tool ideally suited for the visualization
+  of fine structures in extreme ultraviolet images of the corona is presented
+  in this paper. The Noise Adaptive Fuzzy Equalization method is particularly
+  suited for the exceptionally high dynamic range images from the Atmospheric
+  Imaging Assembly instrument on the Solar Dynamics Observatory. This method
+  produces artifact-free images and gives significantly better results than
+  methods based on convolution or Fourier transform which are often used for
+  that purpose.}
+}
+"""
+import argparse
+import multiprocessing
+from multiprocessing import Pool
+
+import sunpy.map
+
+import skimage
+from skimage import filters
+
+import matplotlib.pyplot as plt
+import numpy as np
+import tqdm
+
+
+_PREDEF_PARAMS = {
+    # defined by original program
+    # wavelength    # gamma     # nafe
+    "4500":          (1.0,        .05),
+    "1700":          (2.2,        .10),
+    "1600":          (2.2,        .10),
+    "335":           (3.0,        .20),
+    "304":           (5.5,        .15),
+    "211":           (2.7,        .20),
+    "193":           (2.6,        .20),
+    "171":           (2.6,        .20),
+    "131":           (2.9,        .20),
+    "94":            (3.1,        .20),
+    "def":           (2.2,        .20)
+}
+
+
+def nafe(input_map, imargin=None, gamma=None, omargin=None, nafe=False,
+         nafe_weight=None, roi_p=None, roi_f=None, nproc=None, hbins=None,
+         nrs=15, gb=0.3, lb=0.7, gw=0.3, lw=0.7, n=129,
+         c=[1 for _ in range(12)]):
+    """
+    Enhances image "input_map" via "Noise Adaptive Fuzzy Equalization"
+    method for better visualization. Particularly suited for the exceptionally
+    high dynamic range images from the Atmospheric Imaging Assembly instrument
+    on the Solar Dynamics Observatory. This method produces artifact-free
+    images and gives significantly better results than methods based on
+    convolution or Fourier transform which are often used for that purpose.
+
+    Parameters
+    ----------
+    input_map : `~sunpy.map.sources.sdo.AIAMap`
+        The original input image to be enhanced.
+
+    imargin: `~tuple`
+        (black, white) input image values. Pixels with values < black or
+        > white are set to  0.
+
+    gamma : `~float`
+        The gamma value used for gamma_transform.
+
+    omargin: `~tuple`
+        (black, white) output image values.
+
+    nafe : `~bool`
+        Flag for enabling nafe transform and defaults to false.
```

Comment:
Adopted from original software since NAFE algorithm consists of 2 parts one is gamma correction and the other is the actual nafe (fuzzy histogram) which is computationally intensive hence not enabled (the nafe part) by default. I agree this is actually counter-intuitive and will change to default to true.